Routine leak in TestSTUNTURNConcurrency

```
=== RUN   TestSTUNConcurrency
--- PASS: TestSTUNConcurrency (5.20s)
=== RUN   TestTURNConcurrency
=== RUN   TestTURNConcurrency/UDP_Relay
=== RUN   TestTURNConcurrency/TCP_Relay
turn ERROR: 2020/06/30 01:38:28 error when handling datagram: failed to handle Refresh-request from 127.0.0.1:32878: write udp 127.0.0.1:57202->127.0.0.1:32878: use of closed network connection
=== RUN   TestTURNConcurrency/TLS_Relay
=== RUN   TestTURNConcurrency/DTLS_Relay
turn ERROR: 2020/06/30 01:38:28 error when handling datagram: failed to handle Refresh-request from 127.0.0.1:45410: dtls fatal: use of closed network connection
--- PASS: TestTURNConcurrency (8.03s)
    --- PASS: TestTURNConcurrency/UDP_Relay (0.00s)
    --- PASS: TestTURNConcurrency/TCP_Relay (0.00s)
    --- PASS: TestTURNConcurrency/TLS_Relay (0.01s)
    --- PASS: TestTURNConcurrency/DTLS_Relay (0.01s)
=== RUN   TestSTUNTURNConcurrency
    TestSTUNTURNConcurrency: util.go:38: Unexpected routines on test end: 
        goroutine 725 [select]:
        github.com/pion/turn/v2/internal/client.(*PeriodicTimer).Start.func1(0xa881140, 0xa846a80)
        	/go/pkg/mod/github.com/pion/turn/v2@v2.0.3/internal/client/periodic_timer.go:47 +0xa0
        created by github.com/pion/turn/v2/internal/client.(*PeriodicTimer).Start
        	/go/pkg/mod/github.com/pion/turn/v2@v2.0.3/internal/client/periodic_timer.go:41 +0xab
        
        goroutine 724 [select]:
        github.com/pion/turn/v2/internal/client.(*PeriodicTimer).Start.func1(0xa881110, 0xa846a40)
        	/go/pkg/mod/github.com/pion/turn/v2@v2.0.3/internal/client/periodic_timer.go:47 +0xa0
        created by github.com/pion/turn/v2/internal/client.(*PeriodicTimer).Start
        	/go/pkg/mod/github.com/pion/turn/v2@v2.0.3/internal/client/periodic_timer.go:41 +0xab
--- FAIL: TestSTUNTURNConcurrency (10.40s)
```

--- a/gather.go
+++ b/gather.go
@@ -415,11 +415,14 @@ func (a *Agent) gatherCandidatesRelay(ctx context.Context, urls []*URL) {
 					return locConn.Close()
 				},
 			}
-			candidate, err := NewCandidateRelay(&relayConfig)
-			if err != nil {
+			relayConnClose := func() {
 				if relayConErr := relayConn.Close(); relayConErr != nil {
 					a.log.Warnf("Failed to close relay %v", relayConErr)
 				}
+			}
+			candidate, err := NewCandidateRelay(&relayConfig)
+			if err != nil {
+				relayConnClose()
 
 				client.Close()
 				closeConnAndLog(locConn, a.log, fmt.Sprintf("Failed to create relay candidate: %s %s: %v\n", network, raddr.String(), err))
@@ -427,6 +430,8 @@ func (a *Agent) gatherCandidatesRelay(ctx context.Context, urls []*URL) {
 			}
 
 			if err := a.addCandidate(ctx, candidate, relayConn); err != nil {
+				relayConnClose()
+
 				if closeErr := candidate.close(); closeErr != nil {
 					a.log.Warnf("Failed to close candidate: %v", closeErr)
 				}

--- a/gather_vnet_test.go
+++ b/gather_vnet_test.go
@@ -3,6 +3,7 @@
 package ice
 
 import (
+	"context"
 	"fmt"
 	"net"
 	"testing"
@@ -411,4 +412,48 @@ func TestVNetGatherWithInterfaceFilter(t *testing.T) {
 
 		assert.NoError(t, a.Close())
 	})
+}
+
+func TestVNetGather_TURNConnectionLeak(t *testing.T) {
+	report := test.CheckRoutines(t)
+	defer report()
+
+	turnServerURL := &URL{
+		Scheme:   SchemeTypeTURN,
+		Host:     vnetSTUNServerIP,
+		Port:     vnetSTUNServerPort,
+		Username: "user",
+		Password: "pass",
+		Proto:    ProtoTypeUDP,
+	}
+
+	// buildVNet with a Symmetric NATs for both LANs
+	natType := &vnet.NATType{
+		MappingBehavior:   vnet.EndpointAddrPortDependent,
+		FilteringBehavior: vnet.EndpointAddrPortDependent,
+	}
+	v, err := buildVNet(natType, natType)
+
+	if !assert.NoError(t, err, "should succeed") {
+		return
+	}
+	defer v.close()
+
+	cfg0 := &AgentConfig{
+		Urls: []*URL{
+			turnServerURL,
+		},
+		NetworkTypes:     supportedNetworkTypes,
+		MulticastDNSMode: MulticastDNSModeDisabled,
+		NAT1To1IPs:       []string{vnetGlobalIPA},
+		Net:              v.net0,
+	}
+	aAgent, err := NewAgent(cfg0)
+	if !assert.NoError(t, err, "should succeed") {
+		return
+	}
+
+	aAgent.gatherCandidatesRelay(context.Background(), []*URL{turnServerURL})
+	// Assert relay conn leak on close.
+	assert.NoError(t, aAgent.Close())
 }

